library/std/src/path.rs

```diff
@@ -962,16 +962,50 @@ impl cmp::Eq for Components<'_> {}
 impl<'a> cmp::PartialOrd for Components<'a> {
     #[inline]
     fn partial_cmp(&self, other: &Components<'a>) -> Option<cmp::Ordering> {
-        Iterator::partial_cmp(self.clone(), other.clone())
+        Some(compare_components(self.clone(), other.clone()))
     }
 }
 
 #[stable(feature = "rust1", since = "1.0.0")]
 impl cmp::Ord for Components<'_> {
     #[inline]
     fn cmp(&self, other: &Self) -> cmp::Ordering {
-        Iterator::cmp(self.clone(), other.clone())
+        compare_components(self.clone(), other.clone())
+    }
+}
+
+#[inline]
+fn compare_components(mut left: Components<'_>, mut right: Components<'_>) -> cmp::Ordering {
```

Comment:
This inline doesn't seem well motivated. Inlining all this into all the call sites doesn't seem necessary for the performance improvement and may even be detrimental.

```suggestion
fn compare_components(mut left: Components<'_>, mut right: Components<'_>) -> cmp::Ordering {
```

Reply:
`Path` and `Components` aren't generic, so afaik not having the `#[inline]` attribute would prevent cross-crate inlining. But I'll redo the benchmarks with and without it.

Reply:
@dtolnay microbenchmarks show no difference with or without the annotation. That said I think they're built as part of the same crate, so they would benefit from inlining that downstream users wouldn't.

See also #81281

But I can still remove it if you prefer.

Reply:
ping @dtolnay

Reply:
If you can't demonstrate a performance benefit to having the inline then I advise removing the inline.

This function is an enormous amount of logic (\>250 instructions). In #81281 all of those are way smaller.

Reply:
Ok, removed that commit.